reference/src/client/src/sa_engine_cipher.c

```diff
@@ -0,0 +1,484 @@
+            cipher_##algorithm##_##keysize##_##mode->nid = NID_##algorithm##_##keysize##_##mode; \
+            cipher_##algorithm##_##keysize##_##mode->block_size = BLOCK_SIZE_##algorithm##_##mode; \
+            cipher_##algorithm##_##keysize##_##mode->key_len = SYM_##keysize##_KEY_SIZE; \
+            cipher_##algorithm##_##keysize##_##mode->iv_len = IV_LEN_##algorithm##_##mode; \
+            cipher_##algorithm##_##keysize##_##mode->flags = (cipher_flags); \
+            cipher_##algorithm##_##keysize##_##mode->init = cipher_init; \
+            cipher_##algorithm##_##keysize##_##mode->do_cipher = cipher_do_cipher; \
+            cipher_##algorithm##_##keysize##_##mode->cleanup = cipher_cleanup; \
+            cipher_##algorithm##_##keysize##_##mode->ctx_size = BLOCK_SIZE_##algorithm##_##mode * 2; \
+            cipher_##algorithm##_##keysize##_##mode->set_asn1_parameters = NULL; \
+            cipher_##algorithm##_##keysize##_##mode->get_asn1_parameters = NULL; \
+            cipher_##algorithm##_##keysize##_##mode->ctrl = cipher_ctrl; \
+            cipher_##algorithm##_##keysize##_##mode->app_data = NULL; \
+        } \
+        mtx_unlock(&engine_mutex); \
+        return cipher_##algorithm##_##keysize##_##mode; \
+    }
+
+#endif
+
+typedef struct {
+    sa_key* key;
+    const unsigned char* iv;
+    sa_cipher_mode mode;
+    sa_cipher_algorithm cipher_algorithm;
+    bool padding;
+    sa_crypto_cipher_context cipher_context;
+    uint8_t tag[16];
+} cipher_add_data;
+
+static int cipher_nids[] = {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        NID_chacha20,
+        NID_chacha20_poly1305,
+#endif
+        NID_aes_128_cbc,
+        NID_aes_128_ecb,
+        NID_aes_128_ctr,
+        NID_aes_128_gcm,
+        NID_aes_256_cbc,
+        NID_aes_256_ecb,
+        NID_aes_256_ctr,
+        NID_aes_256_gcm};
+
+static int cipher_nids_num = (sizeof(cipher_nids) / sizeof(cipher_nids[0]));
+
+static sa_cipher_algorithm cipher_get_cipher_algorithm(int nid) {
+    switch (nid) {
+        case NID_aes_128_ecb:
+        case NID_aes_256_ecb:
+            return SA_CIPHER_ALGORITHM_AES_ECB;
+
+        case NID_aes_128_cbc:
+        case NID_aes_256_cbc:
+            return SA_CIPHER_ALGORITHM_AES_CBC;
+
+        case NID_aes_128_ctr:
+        case NID_aes_256_ctr:
+            return SA_CIPHER_ALGORITHM_AES_CTR;
+
+        case NID_aes_128_gcm:
+        case NID_aes_256_gcm:
+            return SA_CIPHER_ALGORITHM_AES_GCM;
+
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+        case NID_chacha20:
+            return SA_CIPHER_ALGORITHM_CHACHA20;
+
+        case NID_chacha20_poly1305:
+            return SA_CIPHER_ALGORITHM_CHACHA20_POLY1305;
+#endif
+
+        default:
+            return UINT32_MAX;
+    }
+}
+
+static int cipher_init(
+        EVP_CIPHER_CTX* cipher_ctx,
+        const unsigned char* key,
+        const unsigned char* iv,
+        int enc) {
+
+    int result = 0;
```

Comment:
Are these result codes part of the spec for openssl engines? Do -1, 0 and 1 have special meaning?

Reply:
They are part of the spec for the engine function (or the user facing api that calls the engine function). Generally, 1 means success or a positive number means success and the size of the output buffer. Generally 0 means failure, but a negative number can also mean failure with a specific reason.